File: native/android-host/reposearchkit/src/main/java/com/example/benchmark/nativeapp/reposearchkit/RepoSearchBridge.kt
package com.example.benchmark.nativeapp.reposearchkit

import java.util.UUID

/** One repository as the search screen reports it to the host. */
data class SearchedRepository(
  val id: Long,
  val fullName: String,
  val stars: Int,
  val language: String?,
)

/** What the search screen reports back. */
sealed interface RepoSearchEvent {
  data class Succeeded(val keyword: String, val repositories: List<SearchedRepository>) :
    RepoSearchEvent

  data class Failed(val keyword: String, val message: String) : RepoSearchEvent
}

/**
 * What the host app asks the search screen to do. The mirror image of
 * [RepoSearchEvent].
 */
sealed interface RepoSearchCommand {
  /**
   * Replaces the keyword on a screen that is already open. The keyword given
   * to [RepoSearchScreen] only reaches the screen while it is being created.
   */
  data class SetKeyword(val keyword: String) : RepoSearchCommand
}

/** The delegate-style callback. */
fun interface RepoSearchListener {
  fun onRepoSearchEvent(event: RepoSearchEvent)
}

/**
 * The host app's end of the channel, delivering events through either a
 * listener or a lambda.
 *
 * It has the same shape as the bridge the other implementations ship with
 * their embedded screen, so the host apps read the same. Here both ends are
 * Kotlin in one process, so values travel as they are: there is no wire format
 * to encode, which is the baseline the other channels are measured against.
 *
 * Main thread only, like the UI on both ends.
 */
class RepoSearchBridge
internal constructor(
  private val channel: RepoSearchChannel,
  private val listener: RepoSearchListener?,
  private val onEvent: ((RepoSearchEvent) -> Unit)?,
) {
  constructor(
    listener: RepoSearchListener? = null,
    onEvent: ((RepoSearchEvent) -> Unit)? = null,
  ) : this(RepoSearchChannel.shared, listener, onEvent)

  private var listenerId: UUID? = null

  fun start() {
    if (listenerId != null) return
    listenerId =
      channel.addEventListener { event ->
        listener?.onRepoSearchEvent(event)
        onEvent?.invoke(event)
      }
  }

  fun stop() {
    listenerId?.let { id ->
      channel.removeEventListener(id)
      listenerId = null
    }
  }

  /** Sends a command to the search screen. It is dropped when no screen is open. */
  fun send(command: RepoSearchCommand) {
    channel.send(command)
  }
}

/** The in-process message bus between the host app and the search screen. */
internal class RepoSearchChannel {
  private val eventListeners = LinkedHashMap<UUID, (RepoSearchEvent) -> Unit>()
  private val commandListeners = LinkedHashMap<UUID, (RepoSearchCommand) -> Unit>()

  fun addEventListener(listener: (RepoSearchEvent) -> Unit): UUID =
    UUID.randomUUID().also { eventListeners[it] = listener }

  fun removeEventListener(id: UUID) {
    eventListeners.remove(id)
  }

  fun post(event: RepoSearchEvent) {
    eventListeners.values.toList().forEach { it(event) }
  }

  fun addCommandListener(listener: (RepoSearchCommand) -> Unit): UUID =
    UUID.randomUUID().also { commandListeners[it] = listener }

  fun removeCommandListener(id: UUID) {
    commandListeners.remove(id)
  }

  fun send(command: RepoSearchCommand) {
    commandListeners.values.toList().forEach { it(command) }
  }

  companion object {
    val shared = RepoSearchChannel()
  }
}

File: native/ios-host/RepoSearchKit/Sources/RepoSearchKit/RepoSearchBridge.swift
import Foundation

/// One repository as the search screen reports it to the host.
public struct SearchedRepository: Identifiable, Hashable, Sendable {
  public let id: Int
  public let fullName: String
  public let stars: Int
  public let language: String?

  public init(id: Int, fullName: String, stars: Int, language: String?) {
    self.id = id
    self.fullName = fullName
    self.stars = stars
    self.language = language
  }

  init(_ repository: Repository) {
    self.init(
      id: repository.id,
      fullName: repository.fullName,
      stars: repository.stars,
      language: repository.language
    )
  }
}

/// What the search screen reports back.
public enum RepoSearchEvent {
  case succeeded(keyword: String, repositories: [SearchedRepository])
  case failed(keyword: String, message: String)
}

/// What the host app asks the search screen to do. The mirror image of
/// `RepoSearchEvent`.
public enum RepoSearchCommand: Equatable {
  /// Replaces the keyword on a screen that is already open. The keyword given
  /// to `RepoSearchView` only reaches the screen while it is being created.
  case setKeyword(String)
}

public protocol RepoSearchBridgeDelegate: AnyObject {
  func repoSearchBridge(_ bridge: RepoSearchBridge, didReceive event: RepoSearchEvent)
}

/// The host app's end of the channel, delivering events through either a
/// delegate or a closure.
///
/// It has the same shape as the bridge the other implementations ship with
/// their embedded screen, so the host apps read the same. Here both ends are
/// Swift in one process, so values travel as they are: there is no wire format
/// to encode, which is the baseline the other channels are measured against.
///
/// Main thread only, like the UI on both ends.
public final class RepoSearchBridge {
  public weak var delegate: RepoSearchBridgeDelegate?
  public var onEvent: ((RepoSearchEvent) -> Void)?

  private let channel: RepoSearchChannel
  private var listenerID: UUID?

  public convenience init(
    delegate: RepoSearchBridgeDelegate? = nil,
    onEvent: ((RepoSearchEvent) -> Void)? = nil
  ) {
    self.init(channel: .shared, delegate: delegate, onEvent: onEvent)
  }

  init(
    channel: RepoSearchChannel,
    delegate: RepoSearchBridgeDelegate? = nil,
    onEvent: ((RepoSearchEvent) -> Void)? = nil
  ) {
    self.channel = channel
    self.delegate = delegate
    self.onEvent = onEvent
  }

  deinit {
    stop()
  }

  public func start() {
    guard listenerID == nil else { return }
    listenerID = channel.addEventListener { [weak self] event in
      guard let self else { return }
      self.delegate?.repoSearchBridge(self, didReceive: event)
      self.onEvent?(event)
    }
  }

  public func stop() {
    guard let listenerID else { return }
    channel.removeEventListener(listenerID)
    self.listenerID = nil
  }

  /// Sends a command to the search screen. It is dropped when no screen is open.
  public func send(_ command: RepoSearchCommand) {
    channel.send(command)
  }
}

/// The in-process message bus between the host app and the search screen.
final class RepoSearchChannel {
  static let shared = RepoSearchChannel()

  private var eventListeners: [UUID: (RepoSearchEvent) -> Void] = [:]
  private var commandListeners: [UUID: (RepoSearchCommand) -> Void] = [:]

  func addEventListener(_ listener: @escaping (RepoSearchEvent) -> Void) -> UUID {
    let id = UUID()
    eventListeners[id] = listener
    return id
  }

  func removeEventListener(_ id: UUID) {
    eventListeners[id] = nil
  }

  func post(_ event: RepoSearchEvent) {
    for listener in eventListeners.values {
      listener(event)
    }
  }

  func addCommandListener(_ listener: @escaping (RepoSearchCommand) -> Void) -> UUID {
    let id = UUID()
    commandListeners[id] = listener
    return id
  }

  func removeCommandListener(_ id: UUID) {
    commandListeners[id] = nil
  }

  func send(_ command: RepoSearchCommand) {
    for listener in commandListeners.values {
      listener(command)
    }
  }
}
